dlls/winevulkan/nis.c

```diff
@@ -0,0 +1,1365 @@
+        return res;
+    }
+
+    allocInfo.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    allocInfo.allocationSize = memory_size;
+    allocInfo.memoryTypeIndex = memory_type;
+
+    res = device->funcs.p_vkAllocateMemory(device->device, &allocInfo, NULL, &nis_data->memory);
+    if (res != VK_SUCCESS){
+        ERR("vkAllocateMemory failed NIS data, res=%d\n", res);
+        return res;
+    }
+
+    res = device->funcs.p_vkBindBufferMemory(device->device, nis_data->uniform_buffer, nis_data->memory, 0);
+    if (res != VK_SUCCESS){
+        ERR("vkBindBufferMemory failed for uniform buffer, res=%d\n", res);
+        return res;
+    }
+
+    res = device->funcs.p_vkBindImageMemory(device->device, nis_data->scaler_coef_img, nis_data->memory, scaler_offset);
+    if (res != VK_SUCCESS){
+        ERR("vkBindImageMemory failed scaler image, res=%d\n", res);
+        return res;
+    }
+
+    res = device->funcs.p_vkBindImageMemory(device->device, nis_data->usm_coef_img, nis_data->memory, usm_offset);
+    if (res != VK_SUCCESS){
+        ERR("vkBindImageMemory failed usm image, res=%d\n", res);
+        return res;
+    }
+
+    return VK_SUCCESS;
+}
+
+static VkResult init(VkDevice device, struct VkSwapchainKHR_T *swapchain)
+{
+    VkResult res;
+    VkSamplerCreateInfo samplerInfo = {0};
+    VkDescriptorPoolSize poolSizes[4] = {{0}, {0}, {0}, {0}};
+    VkDescriptorPoolCreateInfo poolInfo = {0};
+    VkDescriptorSetLayoutBinding layoutBindings[6] = {{0}, {0}, {0}, {0}, {0}, {0}};
+    VkDescriptorSetLayoutCreateInfo descriptorLayoutInfo = {0};
+    VkPipelineLayoutCreateInfo pipelineLayoutInfo = {0};
+    VkShaderModuleCreateInfo shaderInfo = {0};
+    VkShaderModule shaderModule = 0;
+    VkImageCreateInfo imageInfo = {0};
+    VkFormatProperties formatProperties = {0};
+    struct nis_data *nis_data;
+    const char* sharpness;
+    const char *fp16;
+
+    uint32_t i;
+#if defined(USE_STRUCT_CONVERSION)
+    VkImageViewCreateInfo_host viewInfo = {0};
+    VkBufferCreateInfo_host bufferInfo = {0};
+    VkComputePipelineCreateInfo_host pipelineInfo = {0};
+#else
+    VkImageViewCreateInfo viewInfo = {0};
+    VkBufferCreateInfo bufferInfo = {0};
+    VkComputePipelineCreateInfo pipelineInfo = {0};
+#endif
+
+    nis_data = calloc(1, sizeof(struct nis_data));
+    if (!nis_data){
+        WARN("calloc failed\n");
+        return VK_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    swapchain->upscaler_data = nis_data;
+
+    /* Get the NIS upscaler options */
+    fp16 = getenv("WINE_NIS_UPSCALER_USE_FP16");
+    nis_data->use_fp16 = fp16 && strcmp(fp16, "0");
+
+    device->phys_dev->instance->funcs.p_vkGetPhysicalDeviceFormatProperties(device->phys_dev->phys_dev, VK_FORMAT_R16G16B16A16_SFLOAT, &formatProperties);
+    if (!(formatProperties.optimalTilingFeatures & VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT)){
+        WARN("FP16 coefficients are enabled but the device does not support using them as a sampled image, falling back to FP32\n");
```

Comment:
This warning seems to disagree with the logic. Does the condition need `if (use_fp16)`, too?

Reply:
Yes, that should be part of the condition here as well. Good catch.

Reply:
I have made it so FP16 is the default mode if the device supports it and the user has not enabled FP32 usage.